### DISCUSSION TITLE
Is there a way to save variables between functions and requests of same session?

### DISCUSSION THREAD

**Discussion by amks1** [zauberzeug/nicegui#373] [Q&A]
I'm coming from PywebIO running on Tornado, and I'm used to storing variables like: 
```python
pywebio.session.local.foo = bar
```
so that then it's accessible by any modular function that's being called by the entrypoint. 
What I'm trying to achieve is:

```python
from starlette.requests import Request
from nicegui import ui, app
from starlette.middleware.sessions import SessionMiddleware


def modular_func1():
	ui.label(f'I want to access foo here without passing it as a function argument!')


@ui.page('/')
def nicegui_entrypoint(request: Request):
	ui.label(f'Hello World: {request.query_params}')
	foo = 'bar'
	modular_func1()
	

if __name__ == "__main__":
    app.add_middleware(SessionMiddleware, secret_key='SECRET_KEY')
    ui.run(
        host='0.0.0.0',
        port=8000,
        title='My App',
        favicon='',
        show=False,
        reload=False,
    )
```

**Comment by falkoschindler** [comment]
You can define any variable as `global` and read it from anywhere:

```py
def modular_func1():
    ui.label(f'Here is {foo=}')

@ui.page('/')
def nicegui_entrypoint(request: Request):
    ui.label(f'Hello World: {request.query_params}')
    global foo
    foo = 'bar'
    modular_func1()
```

**Reply by amks1** [reply]
> You can define any variable as `global` and read it from anywhere:

I think I oversimplified my code a bit. The functions would be in different files. And in my case these would also be dynamically imported at runtime using `importlib`.

**Reply by falkoschindler** [reply]
Ok, I see. NiceGUI itself uses a strategy of a `nicegui.globals` module that holds one common state that can be imported from all other modules. That's where the `app` instance lives, for example. You could do something similar. (Not to mention various more elaborate approaches using databases.)

**Reply by amks1** [reply]
Here's the implementation in Pywebio:
https://github.com/pywebio/PyWebIO/blob/f0b7beb9c156bca0e973344b6bfe93f911e8328b/pywebio/session/__init__.py#L483

How challenging do you think it would be to implement something like this in NiceGUI?

**Reply by falkoschindler** [reply]
The main purpose of `pywebio.session.local` seems to be to store data not only across multiple functions, but for multiple requests within a _session_. Have look at https://nicegui.io/reference#sessions: It uses FastAPI's SessionMiddleware to achieve exactly this.

**Reply by amks1** [reply]
Unless I'm mistaken, the disadvantage with this is that you'd have to keep passing around the Starlette request object to any functions you're calling.

**Reply by falkoschindler** [reply]
Oh, you're right.

**Reply by rodja** [reply]
The accessibility would still be the same if we incorporate https://github.com/alex-oleshkevich/starsessions or similar.

**Comment by bapowell** [comment]
@amks1, as a word of warning, I've had mixed success using importlib (dynamically importing configurations at runtime) along with a "globals" module, like @falkoschindler mentioned.

For some reason, which I was unable to determine the root cause, the globals module, when used via importlib, seemed to be loaded into a completely different context/namespace than the rest of app, which imported the globals module normally.

**Reply by rodja** [reply]
If globals do not work with importlib, NiceGUI itself would not work because it uses this technique internally quite heavy. Also the PyWebIO implementation mentioned by @amks1 relies on it. Maybe it would be good to identify the root cause of the issues you have been facing...

**Reply by bapowell** [reply]
Sorry, I didn't mean to imply it would never work. A globals module does indeed work, while at the same time using importlib to load configs dynamically. I've done that on a prior project. Rather, I was merely wanting to give a _heads-up_ that it can be tricky, depending on how it's implemented.

Recently I was working on another project (with a very tight deadline) where I ran into a strange edge case doing this. I suspect it had something to do with my use of copious amounts of relative imports... and in general just being in a hurry. It was very frustrating trying to figure out the _bug_ in my program, when it wasn't using the right configuration. I narrowed it down to this use of importlib. Being in a hurry, I implemented some other workaround (don't remember exactly what), and moved on.

I'll see if I can replicate the problem, and present it here. I certainly would appreciate the insight into Python's import mechanisms, especially the relative imports.

**Reply by bapowell** [reply]
I think I found what caused my original problem.

Here was my code:
```python
spec = importlib.util.spec_from_file_location('config', args.configPath)
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)
```
[Here](https://docs.python.org/3.11/library/importlib.html#importing-a-source-file-directly) is what the python docs have to say:
```python
spec = importlib.util.spec_from_file_location(module_name, file_path)
module = importlib.util.module_from_spec(spec)
sys.modules[module_name] = module
spec.loader.exec_module(module)
```
Notice that I was missing that line that inserts the config module into the modules that have been loaded.

**Comment by rodja** [comment]
For sharing information between function calls you can simply use global variables, a database, filesystem or similar as @falkoschindler [wrote above](https://github.com/zauberzeug/nicegui/discussions/373#discussioncomment-4957023). So the remaining question here would be: 

How to make the session object more accessible?

**Reply by falkoschindler** [reply]
That's a good question. See #449 for another example.
How do we proceed? Should we open an issue, or discuss possible APIs first?

**Reply by falkoschindler** [reply]
I'll close this discussion in favor of a new explicit idea/feature request https://github.com/zauberzeug/nicegui/discussions/732.
